### ISSUE TITLE
ValueError: invalid shape in fixed-type tuple: dimension does not fit into a C int.

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#6104]
_Issue [JP-2119](https://jira.stsci.edu/browse/JP-2119) was created on JIRA by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly):_

During a DarkPipeline run on an NRS_DARK exposure, the superbias appears to have a read pattern not listed in the schema enum:

```python
2021-05-31 13:03:05,827 - CRDS - DEBUG -  Reference type 'superbias' defined as 'jwst_nirspec_superbias_0117.fits'
2021-05-31 13:03:05,834 - stpipe.DarkPipeline.superbias - INFO - Using SUPERBIAS reference file /ifs/archive/test/jwst/ref/tmp_crds/cache/references/jwst/nirspec/jwst_nirspec_superbias_0117.fits
2021-05-31 13:03:06,646 - stpipe.DarkPipeline.superbias - WARNING - /dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/stdatamodels/validate.py:34: ValidationWarning: While validating meta.exposure.readpatt the following error occurred:
'ALLIRS2' is not one of ['ACQ1', 'ACQ2', 'BRIGHT1', 'BRIGHT2', 'DEEP2', 'DEEP8', 'FAST', 'FASTGRPAVG', 'FASTGRPAVG8', 'FASTGRPAVG16', 'FASTGRPAVG32', 'FASTGRPAVG64', 'FASTR1', 'FASTR100', 'FGS', 'FGS60', 'FGS8370', 'FGS840', 'FGSRAPID', 'FINEGUIDE', 'ID', 'MEDIUM2', 'MEDIUM8', 'NIS', 'NISRAPID', 'NRS', 'NRSIRS2', 'NRSN16R4', 'NRSN32R8', 'NRSN8R2', 'NRSRAPID', 'NRSIRS2RAPID', 'NRSRAPIDD1', 'NRSRAPIDD2', 'NRSRAPIDD6', 'NRSSLOW', 'RAPID', 'SHALLOW2', 'SHALLOW4', 'SLOW', 'SLOWR1', 'TRACK', 'ANY', 'N/A']

Failed validating 'enum' in schema:
    OrderedDict([('title', 'Readout pattern'),

```

This does not end the pipeline run, but the following error is reported later:

```python
----------------------------------------------------------------------
ERROR RUNNING STEP 'DarkPipeline':
    invalid shape in fixed-type tuple: dimension does not fit into a C
    int.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/bin/strun", line 28, in <module>
    step = Step.from_cmdline(sys.argv[1:])
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/stpipe/step.py", line 173, in from_cmdline
    return cmdline.step_from_cmdline(args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/stpipe/cmdline.py", line 339, in step_from_cmdline
    step.run(*positional)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/stpipe/step.py", line 451, in run
    self.save_model(result, idx=idx)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/stpipe/step.py", line 919, in save_model
    output_path = model.save(
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/stdatamodels/model_base.py", line 511, in save
    self.to_fits(output_path, *args, **kwargs)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/stdatamodels/model_base.py", line 629, in to_fits
    ff.write_to(init, *args, **kwargs)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/asdf/fits_embed.py", line 309, in write_to
    self._update_asdf_extension(
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/asdf/fits_embed.py", line 299, in _update_asdf_extension
    self._hdulist.append(self._create_hdu(buff, use_image_hdu))
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/asdf/fits_embed.py", line 280, in _create_hdu
    column = fits.Column(array=data, format=fmt, name='ASDF_METADATA')
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/astropy/io/fits/column.py", line 663, in __init__
    array = self._convert_to_valid_data_type(array)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-1.2.0.20210524-py38/lib/python3.8/site-packages/astropy/io/fits/column.py", line 1341, in _convert_to_valid_data_type
    dtype = np.dtype(numpy_format).base
ValueError: invalid shape in fixed-type tuple: dimension does not fit into a C int.

```

EDIT: However, this file runs to completion through calwebb_dark in the current dev environment, with asdf version = 2.7.3 and astropy version = 4.2.1. The _uncal file is provided if someone would like to double-check.

**Comment by stscijgbot-jp** [comment]
_Comment by [Rosa Diaz](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rdiaz) on [JIRA](https://jira.stsci.edu/browse/JP-2119?focusedCommentId=531078#comment-531078):_

ALLIRS2 is not a readpatt for NIRSpec and should not be in the enum list. If my memory is not failing me, it was an attempt (in the early days of CRDS for JWST) to set up a wildcard like value for the reference file that needed to cover NRSIRS2, and NRSIRS2RAPID readouts. I think the NIRSpec team wanted to use this wildcard but later on we decided to use rather P_ parameters in CRDS to add all these possible cases as an "or" value rather than a wild card. The RMAP might have been manually set up correctly to cover the two valid IRS2 readouts, such that the file is selected correctly for the data, but the actual file was never changed. In principle, this is one of the design strengths of CRDS; i.e., you don't need to redeliver the reference file to be selected correctly, but rather change the RMAP.

**Comment by jdavies-st** [comment]
The READPATT is a validation warning, but I don't believe that is what is causing the subsequent error and traceback.

This could be a problem related to the same seen in #6028 / [JP-2080](https://jira.stsci.edu/browse/JP-2080) with the checksumming.

I.e. `SuperBiasModel` is a subclass of `ReferenceFileModel`, and as such, should not have an ASDF extension.  Yet the delivered CRDS reffile does because it is old:

```
$ fitsinfo /grp/crds/jwst/references/jwst/jwst_nirspec_superbias_0117.fits
Filename: /grp/crds/jwst/references/jwst/jwst_nirspec_superbias_0117.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU      59   ()      
  1  SCI           1 ImageHDU        12   (2048, 3200)   float32   
  2  DQ            1 ImageHDU        15   (2048, 3200)   int32 (rescales to uint32)   
  3  ERR           1 ImageHDU        12   (2048, 3200)   float32   
  4  DQ_DEF        1 BinTableHDU     20   2R x 4C   [J, J, 40A, 80A]   
  5  ASDF          1 ImageHDU         9   (52434415,)   uint8   
```

And so the `skip_fits_update` might not be working well with it.  Question @stscieisenhamer, does SDP run with `SKIP_FITS_UPDATE` turned on or off?

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2119?focusedCommentId=531127#comment-531127):_

I've updated the title of this issue to reflect the actual error message.

**Comment by stscijgbot-jp** [comment]
_Comment by [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly) on [JIRA](https://jira.stsci.edu/browse/JP-2119?focusedCommentId=531166#comment-531166):_

I've found that I cannot reproduce the error locally, but I am able to reproduce the error on a linux machine (jwcalibdev).

**Comment by stscijgbot-jp** [comment]
_Comment by [James Davies](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jdavies) on [JIRA](https://jira.stsci.edu/browse/JP-2119?focusedCommentId=531274#comment-531274):_

Thanks [Tyler Pauly](https://jira.stsci.edu/secure/ViewProfile.jspa?name=tpauly).  Yeah, gave it a go too, and the file processes through the whole `calwebb_dark` pipeline, but then fails when being written out with the traceback above.  I think this is a pretty serious bug, and it would be good if we got to the bottom of it.

Some cursory investigation in turning off `skip_fits_update` yielded nothing, so I don't think it is related to `ReferenceFileModel` or #&#x2060;6028 / JP-2080.

This data is NIRSpec and it is 

Dropping into `pdb` on failure using

```
strun calwebb_dark jw00654002001_02102_00001_nrs1_uncal.fits --debug
```

yields the following:

```
(Pdb) locals()
{'self': name = 'ASDF_METADATA'; format = '3355451285B', 'array': array([[35, 65, 83, ...,  0,  0,  0]], dtype=uint8), 'format': '3355451285B', 'dims': None, 'numpy_format': '|3355451285u1', 'bzeros': {2: 32768, 4: 2147483648, 8: 9223372036854775808}}
(Pdb) w
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/stpipe/cmdline.py(339)step_from_cmdline()
-> step.run(*positional)
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/stpipe/step.py(451)run()
-> self.save_model(result, idx=idx)
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/stpipe/step.py(919)save_model()
-> output_path = model.save(
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/stdatamodels/model_base.py(511)save()
-> self.to_fits(output_path, *args, **kwargs)
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/stdatamodels/model_base.py(629)to_fits()
-> ff.write_to(init, *args, **kwargs)
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/asdf/fits_embed.py(309)write_to()
-> self._update_asdf_extension(
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/asdf/fits_embed.py(299)_update_asdf_extension()
-> self._hdulist.append(self._create_hdu(buff, use_image_hdu))
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/asdf/fits_embed.py(280)_create_hdu()
-> column = fits.Column(array=data, format=fmt, name='ASDF_METADATA')
  /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/astropy/io/fits/column.py(663)__init__()
-> array = self._convert_to_valid_data_type(array)
> /home/jdavies/miniconda3/envs/jwst/lib/python3.8/site-packages/astropy/io/fits/column.py(1341)_convert_to_valid_data_type()
-> dtype = np.dtype(numpy_format).base
(Pdb) array.shape
(1, 3355451285)
```

So it seems like `asdf` is trying to stuff a huge array of dubious dtype into a column.  Very odd.  Maybe [Edward Slavich](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eslavich) might have a look at this?

**Comment by stscijgbot-jp** [comment]
_Comment by [Edward Slavich](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eslavich) on [JIRA](https://jira.stsci.edu/browse/JP-2119?focusedCommentId=531912#comment-531912):_

spacetelescope/stdatamodels#70

**Comment by stscijgbot-jp** [comment]
_Comment by [Edward Slavich](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eslavich) on [JIRA](https://jira.stsci.edu/browse/JP-2119?focusedCommentId=532185#comment-532185):_

The stdatamodels fix has been released in [0.2.2](https://pypi.org/project/stdatamodels/0.2.2/).
